api/transports.go

```diff
@@ -16,17 +19,31 @@ import (
 // that should work just fine on the raw internet (ie not behind a proxy like nginx etc)
 // it'll also redirect http traffic to it's https route counterpart if port 80 is open
 func StartServer(c *config.API, s *http.Server) error {
+	if !c.Enabled {
+		return nil
+	}
+
+	maAddress := c.Address
+	addr, _ := ma.NewMultiaddr(maAddress)
 	if c.ServeRemoteTraffic {
-		s.Addr = fmt.Sprintf(":%d", c.Port)
-	} else {
-		s.Addr = fmt.Sprintf("%s:%d", LocalHostIP, c.Port)
+		maComponents := ma.Split(addr)
+		if len(maComponents) > 0 {
+			port := maComponents[len(maComponents)-1]
+			maAddress = fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", port)
+			addr, _ = ma.NewMultiaddr(maAddress)
+		} else {
+			return errors.New("failed to parse port from API address")
+		}
 	}
```

Comment:
This bit is interesting, seems there's subtle interaction between the `ServeRemoteTraffic` config value and the supplied multiaddr. Not sure `go-multiaddr-net` will parse raw TCP ports correctly, but this is what I'm hoping will be the case:

| ServeRemoteTraffic | addr | valid? |
| ------ | -------------------- | -------- |
| false | `/ip4/0.0.0.0/tcp/1000` |  ❌ |
| true |  `/ip4/0.0.0.0/tcp/1000` |  ✅ |
| false | `/tcp/1000` | ✅ |
| true | `/tcp/1000` | ❌ |

One could argue that `ServeRemoteTraffic` is made redundant by the multiaddress, but I don't think users should be expected to grasp the subtle difference between `/tcp/1000` and `/ip4/0.0.0.0/tcp/1000`. So we have two options:

1. Consider `ServeRemoteTraffic` authoritative, and adjust the the multiaddr to be valid prior to starting.
2. Throw an error for invalid configuration.

I'm tempted to go with option 1, which seems like what you're doing here. We could make a note to improve this in a follow-up PR.

Reply:
Following up on this, if I set `api.serveRemoteTraffic: true`, `qri connect` blows up:

```
$ QRI_BACKTRACE=1 && qri connect
starting IPFS HTTP API:
API server listening on /ip4/127.0.0.1/tcp/5001

📡  Success! You are now connected to the d.web. Here's your connection details:

peername:	b5
profileID:	QmSyDX5LYTiwQi861F5NAwdHrrnd1iRGsoEvCyzQMUyZ4W
API address:	/ip4/127.0.0.1/tcp/2503
RPC address:	/ip4/127.0.0.1/tcp/2504
IPFS Addresses:
  /ip4/127.0.0.1/tcp/4001/p2p/QmdJQpcmGMkNtKYv5bviZ7z9kPh7uLKSMzqdSbemYs3XpH
  /ip4/192.168.1.11/tcp/4001/p2p/QmdJQpcmGMkNtKYv5bviZ7z9kPh7uLKSMzqdSbemYs3XpH
  /ip4/10.137.14.150/tcp/4001/p2p/QmdJQpcmGMkNtKYv5bviZ7z9kPh7uLKSMzqdSbemYs3XpH
  /ip6/::1/tcp/4001/p2p/QmdJQpcmGMkNtKYv5bviZ7z9kPh7uLKSMzqdSbemYs3XpH
You are running Qri v0.9.9-dev

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x4495948]

goroutine 1 [running]:
github.com/multiformats/go-multiaddr.ForEach(0x0, 0x0, 0xc00031f760)
	/Users/b5/go/pkg/mod/github.com/multiformats/go-multiaddr@v0.2.1/util.go:169 +0x58
github.com/multiformats/go-multiaddr-net.dialArgComponents(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/b5/go/pkg/mod/github.com/multiformats/go-multiaddr-net@v0.1.5/convert.go:173 +0xf7
github.com/multiformats/go-multiaddr-net.DialArgs(0x0, 0x0, 0x400, 0xc0007813e0, 0xc00031f8f0, 0x57aeea0, 0xc0007813e0, 0x60)
	/Users/b5/go/pkg/mod/github.com/multiformats/go-multiaddr-net@v0.1.5/convert.go:127 +0x4d
github.com/multiformats/go-multiaddr-net.Listen(0x0, 0x0, 0x0, 0x0, 0x59dafe0, 0xc0007a2920)
	/Users/b5/go/pkg/mod/github.com/multiformats/go-multiaddr-net@v0.1.5/net.go:278 +0x39
github.com/qri-io/qri/api.StartServer(0xc000407180, 0xc00025e0e0, 0x5a19e80, 0xc00011e180)
	/Users/b5/qri/qri/api/transports.go:39 +0x83
github.com/qri-io/qri/api.Server.Serve(0xc0002fff00, 0x5a19e80, 0xc00011e180, 0xc0002fe500, 0xc00031fcc0)
	/Users/b5/qri/qri/api/api.go:136 +0x605
github.com/qri-io/qri/cmd.(*ConnectOptions).Run(0xc0002808a0, 0x5a3f6a0, 0xc00026a280)
	/Users/b5/qri/qri/cmd/connect.go:102 +0x44
github.com/qri-io/qri/cmd.NewConnectCommand.func1(0xc0000da280, 0x67fc808, 0x0, 0x0, 0x0, 0x0)
	/Users/b5/qri/qri/cmd/connect.go:39 +0x97
github.com/spf13/cobra.(*Command).execute(0xc0000da280, 0x67fc808, 0x0, 0x0, 0xc0000da280, 0x67fc808)
	/Users/b5/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0xc00029e280, 0xc000044758, 0x57af638, 0x5a11240)
	/Users/b5/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/b5/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/qri-io/qri/cmd.Execute()
	/Users/b5/qri/qri/cmd/cmd.go:53 +0x153
main.main()
	/Users/b5/qri/qri/main.go:16 +0x20
```

Reply:
Ok, this turns out to be a relatively deep issue. After learning more about what the go implementation of `multiaddr` looks like, I think we should re-write this file to look like this:

```go
package api

import (
	"fmt"
	"net"
	"net/http"

	ma "github.com/multiformats/go-multiaddr"
	manet "github.com/multiformats/go-multiaddr-net"
	"github.com/qri-io/qri/config"
)

// StartServer interprets info from config to start an API server
func StartServer(c *config.API, s *http.Server) error {
	if !c.Enabled {
		return nil
	}

	addr, err := ma.NewMultiaddr(c.Address)
	if err != nil {
		return err
	}

	var listener net.Listener

	if !c.ServeRemoteTraffic {
		// if we're not serving remote traffic, strip off any address details other
		// than a raw TCP address
		portAddr := "2503"
		ma.ForEach(addr, func(comp ma.Component) bool {
			if comp.Protocol().Code == ma.P_TCP {
				portAddr = comp.Value()
				return false
			}
			return true
		})

		// use a raw listener on a local TCP socket
		listener, err = net.Listen("tcp", fmt.Sprintf(":%s", portAddr))
		if err != nil {
			return err
		}
	} else {
		mal, err := manet.Listen(addr)
		if err != nil {
			return err
		}
		listener = manet.NetListener(mal)
	}

	return s.Serve(listener)
}
```

Reply:
Interesting take and thanks for diving deep on this!